Source/Parallelization/WarpXRegrid.cpp

```diff
@@ -30,7 +43,49 @@ WarpX::LoadBalance ()
             : DistributionMapping::makeKnapSack(*costs[lev], nmax);
         RemakeLevel(lev, t_new[lev], boxArray(lev), newdm);
     }
+    mypc->Redistribute();
+}
+
+void
+WarpX::LoadBalanceHeuristic ()
+{
+    AMREX_ALWAYS_ASSERT(costs_heuristic[0] != nullptr);
+    WarpX::ComputeCostsHeuristic();
 
+    const int nLevels = finestLevel();
+    for (int lev = 0; lev <= nLevels; ++lev)
+    {
+        DistributionMapping newdm;
+        amrex::Vector<long> cost_heuristic((*costs_heuristic[lev]).size());
+#ifdef AMREX_USE_MPI
+        // Parallel reduce the costs_heurisitc
+        amrex::Vector<Real>::iterator it = (*costs_heuristic[lev]).begin();
+        amrex::Real* itAddr = &(*it);
+        ParallelAllReduce::Sum(itAddr,
+                               costs_heuristic[lev]->size(),
+                               ParallelContext::CommunicatorSub());
+#endif
+        // Allow for max number of boxes assigned to a rank with vector weights;
+        // see amrex::DistributionMapping::makeKnapSack (const Vector< Real > &rcost)
+        const amrex::Real cost_heuristic_max = *std::max_element((*costs_heuristic[lev]).begin(),
+                                                                 (*costs_heuristic[lev]).end());
+        const amrex::Real scale = (cost_heuristic_max == 0) ? 1.e9 : 1.e9/cost_heuristic_max;
+
+        for (int i = 0; i < (*costs_heuristic[lev]).size(); ++i)
+        {
+            cost_heuristic[i] = long((*costs_heuristic[lev])[i]*scale) + 1L;
+        }
+
+        const amrex::Real nboxes = costs_heuristic[lev]->size();
+        const amrex::Real nprocs = ParallelContext::NProcsSub();
+        amrex::Real efficiency;
+        const bool do_full_knapsack = true;
+        const int nmax = static_cast<int>(std::ceil(nboxes/nprocs*load_balance_knapsack_factor));
+
+        newdm.KnapSackProcessorMap(cost_heuristic, nprocs, &efficiency, do_full_knapsack, nmax);
+
+        RemakeLevel(lev, t_new[lev], boxArray(lev), newdm);
+    }
     mypc->Redistribute();
 }
```

Comment:
For knapsack, could we simply call
```c++
static DistributionMapping makeKnapSack   (const Vector<Real>& rcost);
```
so that `WarpX:: LoadBalanceHeuristic` is almost identical to `WarpX::LoadBalanceTimers`?

And, @atmyers @WeiqunZhang, could `makeSFC` have the option to take a vector of Real as an input argument, consistent with `makeKnapSack`?

Reply:
If we call this way, `load_balance_knapsack_factor` will have no effect when using the heuristic costs update (the way it is currently written allows `load_balance_knapsack_factor` to work the same way whether using `timers` or `heuristic`).  Ideally there would be a call like `static DistributionMapping makeKnapSack   (const Vector<Real>& rcost, const int nmax)` in AMReX.  Should we leave this as is or use `DistributionMapping makeKnapSack   (const Vector<Real>& rcost)` (and add documentation that `load_balance_knapsack_factor` does not work with heuristic costs update)?

Reply:
@atmyers is working on implementing
```c++
DistributionMapping::makeSFC(*costs[lev], false)
DistributionMapping::makeKnapSack(*costs[lev], nmax);
```
where `costs[lev]` is an `amrex::Vector<amrex::Real>`. Will that work?

Reply:
Yes sounds good, I will add it when the call is available.